### DISCUSSION TITLE
What's the reason to keep inconsistent between createSignal and createStore?

### DISCUSSION THREAD

**Discussion by xuchaoqian** [solidjs/solid#1435] [General]
Let's see two code snippets:

1) Signal
```
const [count, setCount] = createSignal(0); // The first returned value is a getter (a function returning the current value) 
setInterval(() => setCount(count() + 1), 1000);  // Uses the signal: count()
```

2) Store
```
const [todos, setTodos] = createStore([]); // The first element is the store proxy which is readonly
const addTodo = (text) => {
  setTodos([...todos, { id: ++todoId, text, completed: false }]); // Uses the store: todos
};
```
From consistent design perspective, we should use count instead of count().
What's the reason to keep inconsistent between createSignal and createStore?

**Comment by atk** [comment]
Those are significant different approaches to reactive state, so it's only reasonable to expect different handling. Also, the store setter function has a lot more going on than the signal's setter, which you seem to have overlooked.

**Comment by ryansolid** [comment]
Not sure I understand the suggestion. Are you suggesting that count as a variable somehow be reactive? We use a function because it is the only we can track. We could track properties of an object but not a primitive value.

**Comment by xuchaoqian** [comment]
Thank you for your responses @atk @ryansolid 

We are looking for performant web frameworks. Currently, there are two options: `solid+solid start` vs. `svelte+svelte kit`.

After careful evaluation,  we think maybe `solid+solid start` is a better option for us. Even though, `solid start` is still in beta phase，but `svelte kit` already releases v1.0.

The key point is that we feel familiar with solid jsx syntax(we use react+mobx at present) and the design for reactive programming.

Everything goes well, only one thing lets us feel unnatural. We use the variable `count`(-like) created by the equation `const [count, setCount] = useState(0);`(-like) everywhere right now. If migrate to solid, we must remind ourselves time and time again that please call function `count()`(-like) everywhere.

Intuitively, there should be an api: after `const [count, setCount] = createState(0);`,  we can then use `count` instead of `count()`.
And `createState()` can be completely built on `createSignal()`, but provides a kind of syntactic sugar, which will reduce cognition costs and migration resistance from react/jsx market.

How to impl this? One way like this:
Transpiles `const [count, setCount] = createState(0);` to `const [count, setCount] = createSignal(0);` 
Transpiles `count` to `count()`
We don't need modify current concepts but provide a secondary primitive.
 
If solid has the `createState()` api, we will make our own naming conventions like this:
```
const [count, setCount] = createState(0);
const [$count, $setCount] = createSignal(0);
```
Signals can be passed as functions to other components, but States can only be passed as values to other components. Then, we have these corresponding relations: 
```
Signal -> function
State -> variable
Store -> proxy
```

How about this? @ryansolid

**Reply by atk** [reply]
Nobody stops you from creating such a babel plugin yourself. Some people have done similar things, e.g. writing a plugin to allow destructuring: https://github.com/orenelbaum/babel-plugin-solid-undestructure.

**Reply by ryansolid** [reply]
This is particularly awkward to introduce core because it isn't composable. It would need to look specifically for `createState`. We also only compile JSX currently.

You might like:
https://github.com/lxsmnsyc/solid-labels
https://github.com/orenelbaum/babel-plugin-reactivars-solid

**Reply by aifrim** [reply]
While it is important to understand and reduce migrations costs of any kind, this seems to be a minor one. Solid & React, while similar, they do have more important differences between them.

Here is some React code that I have seen everywhere

```jsx
const Component = () => {
   const [state, setState] = useState(true);

   const toggle = () => setState(!state);

   const msg = state ? "True" : "False"

   return <div>
     <button onClick={toggle}>Toggle</button>

     <div>{ msg }</div>
   </div>
}
```
If you re-write this to Solid like so

```jsx
const Component = () => {
   const [state, setState] = createSignal(true);

   const toggle = () => setState(!state());

   const msg = state() ? "True" : "False";

   return <div>
     <button onClick={toggle}>Toggle</button>

     <div>{ msg }</div>  {/* this will always render True, even if the state is false */}
   </div>
}
```
Then, it will not work. Since in Solid, the function is executed once, not after every "state" update like in React.

You'd have to do it like this in order for it to work
```jsx
const Component = () => {
   const [state, setState] = createSignal(true);

   const toggle = () => setState(!state());

   return <div>
     <button onClick={toggle}>Toggle</button>

     <div>{ state() ? "True" : "False" }</div>
   </div>
}
```

In React the Components' function is called
- at mount
- at props change
- at state change
- at hook change

Where as in Solid, the Components' is called **only once**. Afterwards all of the changes that happen are managed by Signal. Add to this the onMount, onCleanup, onError effects. The way that Signal manages refs, cleanups...

Due to this paradigm shift, I'd say that this "inconsistency" between the signatures of signals & stores is trivial.

Further more, I'd argue that due to this inherent difference, they are are more easely identified when the developer looks at the code.

**Reply by xuchaoqian** [reply]
> This is particularly awkward to introduce core because` it isn't composable`

@ryansolid 
Would you like to explain more detail about this?  Why to say `it isn't composable`? It will be much helpful to know the traps to use createState() before implementing the custom babel plugin.

**Reply by atk** [reply]
You could easily make your example reactive by wrapping msg in a function:

```js
const msg = () => state() ? "True" : "False";
```

Outside of {} in JSX, you need to opt in to reactivity with solid, opposite to react, where you need to opt out of it to increase performance.

**Reply by orenelbaum** [reply]
@xuchaoqian I think that Ryan is talking about how with these kind of syntaxes you usually have to opt out of the syntax in order to pass around the getter or the setter (or both), since a normal reference to the state will give you the value rather than the getter or setter.

For example let's say you have a function like this:
```ts
const doubleSig = (sig: () => number) => () => sig() * 2;
```
With normal Solid you'd use it this way:
```js
const [count, setCount] = createSignal(0)
const doubleCount = doubleSig(count)
createEffect(() =>
  console.log(() => doubleCount())
)
```

Now let's say you're using [Solid Labels](https://github.com/lxsmnsyc/solid-labels). You can define state like this:
```js
let count = $signal(0)
```
If you try to pass it to `doubleSig` the same way, i.e.
```js
const doubleCount = doubleSig(count)
createEffect(() =>
  console.log(() => doubleCount)
)
```
it's not going to work. It will compile roughly to
```js
const [count, setCount] = createSignal(0)
const doubleCount = doubleSig(count())
createEffect(() =>
  console.log(() => doubleCount)
)
```
There are two problems: we are passing the value of `count` instead of the getter, and `doubleCount` is just a normal variable and not a reactive variable (in the Svelte like syntax meaning of the term). `doubleCount` is a variable that references a reactive signal but it doesn't use the special syntax. That's why when we are referencing it in `createEffect`, the Babel plugin doesn't make it a function call, and thus we are logging the getter instead of the value.

This is how we can fix it:
```js
const doubleCount = $derefMemo(doubleSig($get(count)));
createEffect(() =>
  console.log(() => doubleCount)
)
```
We are using `$get` to opt out of the special syntax and get the actual getter instead of the value. We could achieve the same thing with `() => count`. Then when we get a new getter back from `doubleSig` we pass it into `$derefMemo` to opt back into the special syntax and make `doubleCount` a reactive variable, so that we can reference it later without calling it and the Babel plugin turn it into a function call.
This will compile roughly to the same code as the first snippet with normal Solid.

So you can still pass getters and setters around which is necessary for composition, but it means you have to opt out and opt back in whenever you want to do so, which means that when you're composing state the syntax might have diminishing returns if the goal is to make your code cleaner and more concise.

You can try it here:
https://codesandbox.io/s/cool-wildflower-3gq3hz?file=/src/App.tsx:119-174

With my plugin [Reactivars](https://github.com/orenelbaum/babel-plugin-reactivars-solid) (which works even though it's still a bit WIP), I tried to address this with `$` prefixes on reactive variables.
We can define state this way:
```
let $count = 0
```
If we define `doubleSig` this way
```
const doubleSig = ({ $sig }: { $sig: number }) => () => $sig * 2
```
now we can pass `$count` like this
```
const doubleCount = doubleSig({ $sig: $count })
createEffect(() =>
  console.log(() => doubleCount())
)
```
And it will work. We still had to opt out, but the way we did it is by assigning `$count` to an object property prefixed with `$` (`$sig`), which is a special syntax to opt out.
Also note that `doubleCount` is a normal variable and not a reactive variable, so we didn't opt back in to the syntax. The plugin currently doesn't have a special syntax for getters, only for getter setter pairs.

You can try it here:
https://stackblitz.com/edit/github-1g8eax?file=src/App.tsx

This is not ideal, but I personally prefer this approach, it has some advantages I didn't show here, and like I said it's still WIP, it's hard to come up with a good syntax while also staying compatible with TS and not building a custom language server like Svelte did, which is a much bigger project, but I might in the future make a language server so I can unlock the possibility of a better syntax.

Either way this is a pretty complicated problem to solve properly. If you're satisfied with those reactive variables being local to the scope they are defined in, either of those plugins should work but you might prefer Solid Labels since it supports memos as well.

**Reply by xuchaoqian** [reply]
> @xuchaoqian I think that Ryan is talking about how with these kind of syntaxes you usually have to opt out of the syntax in order to pass around the getter or the setter (or both), since a normal reference to the state will give you the value rather than the getter or setter......

@orenelbaum Thank you for your excellent explanation🎉🎉🎉

Indeed, opt-in & opt-out make things become complicated quickly. 

My first thought was don't use any special syntax for opt-in & opt-out. For example:
```
const [count, setCount] = createState(0)
const doubleCount = count * 2
createEffect(()=>{ console.log(doubleCount); });
```
will compile to:
```
const [count, setCount] = createSignal(0)
const doubleCount = () => count()* 2
createEffect(()=>{ console.log(doubleCount()); });
```

If we need opt-in & opt-out, we just use signals. For example:
 
```
function $doubleCount($count) { return $count() * 2; }
const [$count, $setCount] = createSignal(0)
createEffect(()=>{ console.log($doubleCount()); });
```

We can also use states and signals together. For example:
```
const [hours, setHours] = createState(0)
const [$days, $setDays] = createSignal(0)
const $total() = () => hours * $days()
createEffect(()=>{ console.log($total()); });
```
will compile to:
```
const [hours, setHours] = createState(0)
const [$days, $setDays] = createSignal(0)
const $total() = () => hours() * $days()
createEffect(()=>{ console.log($total()); });
```

The difference between State and Signal is: 
Signals('s getters and setters) can be passed as functions to other components/functions, but states can only be passed as values to other components/functions.

In a sense,  State is a restricted version of Signal, but cannot escape its enclosing scope.

**Reply by orenelbaum** [reply]
This is similar to Svelte with local state and stores. There are some drawbacks to this syntax but if you're fine with it then why not.

**Comment by xuchaoqian** [comment]
Thank you all, I've learned a lot. @atk @ryansolid @aifrim @orenelbaum
